platform/linux-generic/pktio/mdev.c

```diff
@@ -0,0 +1,585 @@
+	     (struct vfio_info_cap_header *)((char *)ptr + info->cap_offset);
+	     hdr != ptr;
+	     hdr = (struct vfio_info_cap_header *)((char *)ptr + hdr->next)) {
+		if (hdr->id == id)
+			return hdr;
+	}
+
+	return NULL;
+}
+
+static struct vfio_info_cap_header *
+vfio_get_cap_info(struct vfio_region_info *region_info, __u16 id)
+{
+	struct vfio_info_cap_header *caps = NULL;
+
+	caps = vfio_get_region_info_cap(region_info, id);
+
+	return caps;
+}
+
+int vfio_get_region_sparse_mmaps(struct vfio_region_info *region_info,
+				 struct vfio_region_info_cap_sparse_mmap
+				 **sparse)
+{
+	struct vfio_info_cap_header *caps = NULL;
+	int ret = -ENOENT;
+
+	if (region_info->flags & VFIO_REGION_INFO_FLAG_CAPS &&
+	    region_info->argsz > sizeof(*region_info)) {
+		caps = vfio_get_cap_info(region_info,
+					 VFIO_REGION_INFO_CAP_SPARSE_MMAP);
+		if (!caps)
+			goto out;
+		vfio_find_sparse_mmaps(caps, sparse);
+		if (*sparse) {
+			for (uint32_t i = 0; i < (*sparse)->nr_areas; i++)
+				ODP_DBG("Sparse region: %d 0x%llx %llu\n", i,
+					(*sparse)->areas[i].offset,
+					(*sparse)->areas[i].size);
+		}
+
+		ret = 0;
+	}
+
+out:
+	return ret;
+}
+
+/** Match capability type
+ * returns 0 on succcess
+ */
+int vfio_get_region_cap_type(struct vfio_region_info *region_info,
+			     mdev_region_class_t *class_info)
+{
+	struct vfio_info_cap_header *caps = NULL;
+	int ret = 0;
+	struct vfio_region_info_cap_type *cap_type;
+
+	if (region_info->flags & VFIO_REGION_INFO_FLAG_CAPS &&
+	    region_info->argsz > sizeof(*region_info)) {
+		caps = vfio_get_cap_info(region_info,
+					 VFIO_REGION_INFO_CAP_TYPE);
+		if (!caps) {
+			ret = -EINVAL;
+			goto out;
+		}
+
+		cap_type =
+		    odp_container_of(caps, struct vfio_region_info_cap_type,
+				     header);
+
+		class_info->type = cap_type->type;
+		class_info->subtype = cap_type->subtype;
+	}
+out:
+	return ret;
+}
+
+/**
+ * Get specific region info
+ */
+static struct vfio_region_info *vfio_get_region(mdev_device_t *mdev,
+						__u32 region)
+{
+	int ret;
+	struct vfio_region_info *region_info = NULL;
+
+	ODP_DBG("Region:%d\n", region);
+	region_info = calloc(1, sizeof(*region_info));
+	if (!region_info)
+		goto out;
+
+	region_info->index = region;
+	region_info->argsz = sizeof(*region_info);
+	ret = ioctl(mdev->device, VFIO_DEVICE_GET_REGION_INFO, region_info);
+	if (ret < 0) {
+		ODP_ERR("Failed to get PCI region info\n");
+		goto out;
+	}
+
+	if (!region_info->size) {
+		ODP_DBG("region info %d is empty, skipping\n", region);
+		goto out;
+	}
+
+	if (region_info->argsz > sizeof(*region_info)) {
+		struct vfio_region_info *tmp;
+
+		tmp = realloc(region_info, region_info->argsz);
+		if (!tmp)
+			goto out;
+
+		region_info = tmp;
+
+		ODP_DBG("region info %d with extended capabilities size: %u\n",
+			region, region_info->argsz);
+		ret = ioctl(mdev->device, VFIO_DEVICE_GET_REGION_INFO,
+			    region_info);
+		if (ret < 0 || !region_info->size) {
+			ODP_ERR("Failed to get PCI region info\n");
```

Comment:
This is inconsistent with the earlier tests above. `ret < 0` above gives `ODP_ERR()` while `!region_info->size` gets an `ODP_DBG()`. It seems `ODP_DBG()` should be used consistently here.

Reply:
This effectively maps Rx/Tx descriptors and a failure indicates that the card will not be operational, that's why it's under ODR_ERR

Reply:
I agree with Bill in some sense here. It indeed needs to be consistent, but instead of making it ODP_DBG I'll just reorder region->size check.